src/coreComponents/codingUtilities/Utilities.hpp

```diff
@@ -142,19 +142,19 @@ VAL findOption( mapBase< KEY, VAL, SORTED > const & map,
   return iter->second;
 }
 
-/**
- * @brief Construct a vector of map keys.
- * @tparam KEY map key type
- * @tparam VAL map value type
- * @tparam SORTED whether map is ordered or unordered
- * @param map the map
- * @return a vector of keys
- */
-template< typename KEY, typename VAL, typename SORTED >
-std::vector< KEY > mapKeys( mapBase< KEY, VAL, SORTED > const & map )
+ /**
+  * @brief Extract the keys from the given map.
+  * @tparam MAP Type of the considered map.
+  * @tparam R Type of the container holding the keys.
+  * @param[in] map The map from which keys will be extracted.
+  * @return The container with the keys.
+  */
+template< typename MAP, typename R = std::vector< typename MAP::key_type > >
+R mapKeys( MAP const & map )
 {
-  std::vector< KEY > keys;
-  std::transform( map.begin(), map.end(), std::back_inserter( keys ), [=]( auto const & p ){ return p.first; } );
+  R keys;
+  std::transform( map.begin(), map.end(), std::inserter( keys, keys.end() ),
+                  [=]( auto const & p ) { return p.first; } );
```

Comment:
Hi @klevzoff I somehow needed to be able to get the keys as a `set` so I ended up with this implementation.

But I'm not happy with this implementation since optional `R` depends on `MAP` because of the nested type `key_type` (and optional parameters go at the end).

So when I call with anything different than `std::vector< typename MAP::key_type >` I need to define `MAP` in the template call. I get like `mapKeys< decltype( m ), std::set< string > >( m )` which upsets me.

Do you have any idea?

Reply:
How about making it a template template parameter? :smile: 
```c++
template< template< typename > class R, typename MAP >
R< typename MAP::key_type > mapKeys( MAP const & map )
{
  R< typename MAP::key_type > keys;
  ...
}
...
auto keyVec = mapKeys< std::vector >( map );
auto keySet = mapKeys< std::set >( map );
```
(I haven't tested this code)

Reply:
I'm fine with something like what you proposed, but that means that there is no more default.
I'm OK with that nevertheless...

Reply:
> I did not know that we could provide "incomplete types" like `std::vector`! Good!

Yes, I keep forgetting it too and get surprised by this fact from time to time :laughing:

Reply:
Hmm, it appears that this does not work with `-std=c++14` 😢

Reply:
Hmm, it can be worked around by adding
```c++
template< typename T >
using set = std::set< T >;

template< typename T >
using vector = std::vector< T >;
```
and then using those as parameters. But it's not as nice :disappointed: .

Reply:
Alternative solution (note the added `...`):
```c++
template< template< typename ... > class R = std::vector, typename MAP >
R< typename MAP::key_type > mapKeys( MAP const & map )
{
  R< typename MAP::key_type > keys;
  std::transform( map.begin(), map.end(), std::inserter( keys, keys.end() ),
                  [=]( auto const & p ) { return p.first; } );
  return keys;
}
```

Reply:
Brilliant! I've just pushed the patch!